File: postInTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
idx = 0
maps = {}
class Solution:
   
    def fetchTree(self,inorder, preorder, start, end ):
        global idx, maps
        if start > end:# or idx > len(preorder):
            return None
        rootval = preorder[idx]
        idx = idx + 1
        root = TreeNode(rootval) 
        rootidx = maps.get(rootval)  
        root.left =  self.fetchTree(inorder, preorder, start, rootidx - 1,)
        root.right = self.fetchTree(inorder, preorder, rootidx + 1 , end)
        return root
    
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        global maps
        global idx
        idx = 0
        maps = {}
        if inorder is None or len(inorder) < 1:
            return None
        # if len(inorder) == 1:
        #     return TreeNode(inorder[0]) 
        i = 0
        while( i < len(inorder)):
            maps[inorder[i]] = i
            i += 1
        return self.fetchTree(inorder, preorder, 0 , len(preorder) - 1)
   

File: src/PostInTree.java
import java.util.HashMap;
import java.util.Map;
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no



public class PostInTree {
    /**
     * Definition for a binary tree node.
     * public class TreeNode {
     *     int val;
     *     TreeNode left;
     *     TreeNode right;
     *     TreeNode() {}
     *     TreeNode(int val) { this.val = val; }
     *     TreeNode(int val, TreeNode left, TreeNode right) {
     *         this.val = val;
     *         this.left = left;
     *         this.right = right;
     *     }
     * }
     */
    class Solution {
        int ind ;
        Map<Integer, Integer> map;
        public class TreeNode {
            int val;
            TreeNode left;
            TreeNode right;
            TreeNode() {}
            TreeNode(int val) { this.val = val; }
            TreeNode(int val, TreeNode left, TreeNode right) {
                this.val = val;
                this.left = left;
                this.right = right;
            }
        }

        public TreeNode buildTree(int[] inorder, int[] postorder) {
            ind = postorder.length -1;
            map = new HashMap<>();
            if(postorder == null ||postorder.length ==0 || inorder.length ==0) return null;
            for(int i=0; i < inorder.length;i++) {
                map.put(inorder[i], i);
            }
            TreeNode root = helper(postorder, inorder, 0, postorder.length-1);

            return root;
        }
        private TreeNode helper(int[] postorder, int[] inorder, int start, int end ){
            if(start > end){
                return  null ;
            }
            int rootinx = postorder[ind];
            TreeNode root = new TreeNode(rootinx);
            ind--;
            int rootIdx = map.get(root.val);
            TreeNode right = helper(postorder, inorder, rootIdx+1, end);
            TreeNode left = helper(postorder, inorder, start, rootIdx-1);

            root.left = left;
            root.right = right;
            return root;
        }
    }

}
